src/lib.rs

```diff
@@ -12,99 +12,155 @@
-    match termios::tcsetattr(STDIN_FILENO, termios::TCSANOW, &term_orig) {
-        Ok(_) => {},
-        Err(err) => {
-            unsafe { password.as_mut_vec() }.set_memory(0);
-            return Err(err);
-        }
+        Ok(password)
     }
 
-    // Remove the \n from the line.
-    password.pop().unwrap();
-
-    Ok(password)
+    #[test]
+    fn it_works() {
+        let term_before = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(read_password().unwrap().as_slice(), "my-secret");
+        let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(term_before, term_after);
+        unsafe { TEST_EOF = true; }
+        assert!(!read_password().is_ok());
+        let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
+        assert_eq!(term_before, term_after);
+        assert!(unsafe { TEST_HAS_SEEN_REGULAR_BUFFER });
+        assert!(unsafe { TEST_HAS_SEEN_EOF_BUFFER });
+    }
 }
-
-#[test]
-fn it_works() {
-    let term_before = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(read_password().unwrap().as_slice(), "my-secret");
-    let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(term_before, term_after);
-    unsafe { TEST_EOF = true; }
-    assert!(!read_password().is_ok());
-    let term_after = termios::Termios::from_fd(STDIN_FILENO).unwrap();
-    assert_eq!(term_before, term_after);
-    assert!(unsafe { TEST_HAS_SEEN_REGULAR_BUFFER });
-    assert!(unsafe { TEST_HAS_SEEN_EOF_BUFFER });
+#[cfg(windows)]
+mod windows {
+    extern crate winapi;
+    extern crate "kernel32-sys" as kernel32;
+    use std::old_io::{IoError, IoErrorKind, IoResult};
+    use std::ptr::null_mut;
+    pub fn read_password() -> IoResult<String> {
+        unsafe {
+            let handle = kernel32::GetStdHandle(winapi::STD_INPUT_HANDLE);
+            if handle == winapi::INVALID_HANDLE_VALUE {
+                return Err(IoError::last_error())
+            }
+            let mut mode = 0;
+            if kernel32::GetConsoleMode(handle, &mut mode as winapi::LPDWORD) == 0 {
+                return Err(IoError::last_error())
+            }
+            if kernel32::SetConsoleMode(
+                handle, winapi::ENABLE_LINE_INPUT | winapi::ENABLE_PROCESSED_INPUT,
+            ) == 0 {
+                return Err(IoError::last_error())
+            }
+            let mut buf: [winapi::WCHAR; 0x1000] = [0; 0x1000];
+            let mut read = 0;
+            if kernel32::ReadConsoleW(
+                handle, buf.as_mut_ptr() as winapi::LPVOID, 0x1000,
+                &mut read as winapi::LPDWORD, null_mut(),
+            ) == 0 {
+                let err = IoError::last_error();
+                kernel32::SetConsoleMode(handle, mode);
+                return Err(err)
+            }
+            if kernel32::SetConsoleMode(handle, mode) == 0 {
+                return Err(IoError::last_error())
+            }
+            println!("");
```

Comment:
Is this because in Windows, the `\r\n` is not displayed at all, so we need to display it manually? If so, it would be nice to add a comment about it. This is not obvious for me as a UNIX user.